File: src/main/java/de/fhopf/akka/actor/Master.java
package de.fhopf.akka.actor;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import akka.actor.ActorRef;
import akka.actor.UntypedActor;
import de.fhopf.akka.PageContent;
import de.fhopf.akka.VisitedPageStore;


/**
 * Master actor that coordinates the two actors.
 *
 * @author Florian Hopf, http://www.florian-hopf.de
 */
public abstract class Master extends UntypedActor {

    private final Logger logger = LoggerFactory.getLogger(Master.class);
    private final VisitedPageStore visitedPageStore = new VisitedPageStore();

    @Override
    public void onReceive(Object message) throws Exception {

        if (message instanceof String) {
            // start
            String start = (String) message;
            visitedPageStore.add(start);
            getParser().tell(visitedPageStore.getNext(), getSelf());
        } else if (message instanceof PageContent) {
            PageContent content = (PageContent) message;
            getIndexer().tell(content, getSelf());
            visitedPageStore.addAll(content.getLinksToFollow());

            logger.info(visitedPageStore.toString());

            if (visitedPageStore.isFinished()) {
                getIndexer().tell(IndexingActor.COMMIT_MESSAGE, getSelf());
            } else {
                for (String page : visitedPageStore.getNextBatch()) {
                    getParser().tell(page, getSelf());
                }
            }
        } else if (message instanceof IndexedMessage) {
            IndexedMessage indexedMessage = (IndexedMessage) message;
            visitedPageStore.finished(indexedMessage.path);

            if (visitedPageStore.isFinished()) {
                getIndexer().tell(IndexingActor.COMMIT_MESSAGE, getSelf());
            }
        } else if (message == IndexingActor.COMMITTED_MESSAGE) {
            logger.info("Shutting down, finished");
            getContext().system().shutdown();
        }
    }


    protected abstract ActorRef getIndexer();


    protected abstract ActorRef getParser();
}

File: src/main/scala/de/fhopf/akka/actor/Master.scala
package de.fhopf.akka.actor

import akka.actor.ActorRef
import akka.actor.UntypedActor
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import java.util.concurrent.CountDownLatch
import de.fhopf.akka._

/**
 * @author Florian Hopf, http://www.florian-hopf.de
 * @author Alfredo Serafini, http://www.seralf.it
 */
abstract class Master(latch: CountDownLatch) extends UntypedActor {

  private var logger: Logger = LoggerFactory.getLogger(classOf[Master])
  private var visitedPageStore: VisitedPageStore = new VisitedPageStore()

  override def onReceive(message: Any) = {

    message match {
      case start: String => {
        visitedPageStore.add(start)
        getParser().tell(visitedPageStore.getNext(), getSelf())
      }
      case content: PageContent => {
        getIndexer().tell(content, getSelf())
        visitedPageStore.addAll(content.getLinksToFollow())
        if (visitedPageStore.isFinished()) {
          getIndexer().tell(new COMMIT_MESSAGE(), getSelf())
        } else {
          // TODO
          //				for (String page : visitedPageStore.getNextBatch()) {
          //					getParser().tell(page, getSelf())
          //				}
        }

      }
      case indexedMessage: IndexedMessage => {
        visitedPageStore.finished(indexedMessage.getPath)
        if (visitedPageStore.isFinished())
          getIndexer().tell(new COMMIT_MESSAGE(), getSelf())

      }
      case _: COMMITTED_MESSAGE => {
        logger.info("Shutting down, finished")
        getContext().system.shutdown()
        latch.countDown()
      }
    }

  }

  protected def getIndexer(): ActorRef

  protected def getParser(): ActorRef
}
